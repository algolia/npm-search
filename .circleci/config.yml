version: 2.1

orbs:
  node: circleci/node@5.1.0

aliases:
  # Forward the current folder when using workflows
  - &persist_work_dir
    root: .
    paths:
      - .

  - &attach_work_dir
    at: ~/repo

jobs:
  # Step build
  build:
    docker:
      - image: cimg/node:16.17.0

    working_directory: ~/repo

    steps:
      - checkout
      - persist_to_workspace: *persist_work_dir

      - node/install-packages:
          pkg-manager: yarn

      - run:
          name: Lint
          command: yarn lint

      - run:
          name: Build
          command: yarn build

      - run:
          name: Test
          command: yarn test

      - persist_to_workspace: *persist_work_dir
      - store_test_results:
          path: junit/

  # Step release
  release:
    docker:
      - image: cimg/node:16.17.0
    working_directory: ~/repo
    steps:
      - attach_workspace: *attach_work_dir

      - setup_remote_docker:
          version: 19.03.13

      - run:
          name: Release
          command: yarn semantic-release
      - run:
          name: Docker Build
          command: yarn docker:build
      - run:
          name: Docker publish
          command: |
            echo "$DOCKERHUB_PASS" | docker login -u "$DOCKERHUB_USERNAME" --password-stdin
            yarn docker:release

workflows:
  suite:
    jobs:
      - build

      - release:
          requires:
            - build
          filters:
            branches:
              only: master
