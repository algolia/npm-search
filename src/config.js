import { config } from 'dotenv';
config();

const defaultConfig = {
  npmRegistryEndpoint: 'https://replicate.npmjs.com/registry',
  npmDownloadsEndpoint: 'https://api.npmjs.org/downloads',
  maxObjSize: 450000,
  popularDownloadsRatio: 0.005,
  appId: 'OFCNCOG2CU',
  apiKey: '',
  indexName: 'npm-search',
  bootstrapIndexName: 'npm-search-bootstrap',
  replicateConcurrency: 10,
  bootstrapConcurrency: 100,
  timeToRedoBootstrap: 7 * 24 * 3600 * 1000 /* one week */,
  seq: null,
  indexSettings: {
    searchableAttributes: [
      'unordered(popularName)',
      'unordered(name)',
      'unordered(concatenatedName)',
      'unordered(description)',
      'unordered(keywords)',
      'owner.name',
      'owners.name',
    ],
    attributesForFaceting: [
      'filterOnly(concatenatedName)' /* optionalFacetFilters to boost the name */,
      'searchable(keywords)',
      'searchable(owner.name)',
      'deprecated',
    ],
    customRanking: ['desc(downloadsLast30Days)'],
    disablePrefixOnAttributes: ['keywords', 'owner.name', 'owners.name'],
    disableExactOnAttributes: [
      'description',
      'keywords',
      'owner.name',
      'owners.name',
    ],
    disableTypoToleranceOnAttributes: ['keywords'],
    exactOnSingleWordQuery: 'attribute',
    ranking: [
      'filters',
      'typo',
      'words',
      'proximity',
      'attribute',
      'asc(deprecated)',
      'asc(badPackage)',
      'desc(popular)',
      'exact',
      'custom',
    ],
    optionalWords: ['js', 'javascript'],
    separatorsToIndex: '_',
    replaceSynonymsInHighlight: false,
  },
  indexSynonyms: [
    {
      type: 'synonym',
      synonyms: ['_', 'underscore'],
      objectID: 'underscore',
    },
  ],
  indexRules: [
    {
      objectID: 'promote-exact',
      description: 'promote exact matches',
      condition: {
        pattern: '{facet:concatenatedName}',
        anchoring: 'is',
      },
      consequence: {
        params: {
          automaticOptionalFacetFilters: ['concatenatedName'],
        },
      },
    },
  ],
};

export default Object.entries(defaultConfig).reduce(
  (res, [key, defaultValue]) => ({
    ...res,
    [key]:
      key in process.env
        ? JSON.parse(
            typeof defaultValue === 'string'
              ? `"${process.env[key]}"`
              : process.env[key]
          )
        : defaultValue,
  }),
  {}
);
