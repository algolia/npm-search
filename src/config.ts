import type { Settings, Synonym, Rule } from '@algolia/client-search';
import ms from 'ms';

const indexSettings: Settings = {
  searchableAttributes: [
    'unordered(_searchInternal.popularName)',
    'name, description, keywords',
    '_searchInternal.alternativeNames',
    'owner.name',
    'owners.name',
  ],
  attributesForFaceting: [
    'filterOnly(_searchInternal.alternativeNames)' /* optionalFacetFilters to boost the name */,
    'filterOnly(bin)',
    'searchable(keywords)',
    'searchable(computedKeywords)',
    'searchable(owner.name)',
    '_searchInternal.expiresAt',
    'deprecated',
    'isDeprecated',
    'types.ts',
    'moduleTypes',
    'styleTypes',
    'popular',
  ],
  customRanking: [
    'desc(_searchInternal.downloadsMagnitude)',
    'desc(_searchInternal.jsDelivrPopularity)',
    'desc(dependents)',
    'desc(downloadsLast30Days)',
  ],
  disablePrefixOnAttributes: ['owner.name', 'owners.name'],
  disableExactOnAttributes: ['owner.name', 'owners.name'],
  exactOnSingleWordQuery: 'attribute',
  ranking: [
    'filters',
    'typo',
    'words',
    'proximity',
    'attribute',
    'asc(deprecated)',
    'asc(isDeprecated)',
    'asc(badPackage)',
    'desc(popular)',
    'exact',
    'custom',
  ],
  optionalWords: ['js', 'javascript'],
  separatorsToIndex: '_',
  replaceSynonymsInHighlight: false,
  maxValuesPerFacet: 1000,
};

const indexSynonyms: Synonym[] = [
  {
    type: 'synonym',
    synonyms: ['_', 'underscore'],
    objectID: 'underscore',
  },
  {
    type: 'synonym',
    synonyms: ['a11y', 'accessibility', 'accessible'],
    objectID: 'a11y',
  },
  {
    type: 'synonym',
    synonyms: [
      'i18n',
      'internationalisation',
      'internationalization',
      'translation',
      'translate',
    ],
    objectID: 'i18n',
  },
  {
    type: 'synonym',
    synonyms: ['k8s', 'kubernetes'],
    objectID: 'k8s',
  },
];

const indexRules: Rule[] = [
  {
    objectID: 'promote-exact',
    description: 'promote exact matches',
    condition: {
      pattern: '{facet:_searchInternal.alternativeNames}',
      anchoring: 'is',
    },
    consequence: {
      params: {
        automaticOptionalFacetFilters: [
          {
            facet: '_searchInternal.alternativeNames',
          },
        ],
      },
    },
  },
  {
    condition: {
      pattern: 'author\\: {facet:owner.name}',
      anchoring: 'contains',
    },
    consequence: {
      params: {
        automaticFacetFilters: ['owner.name'],
        query: {
          remove: ['author\\:', '{facet:owner.name}'],
        },
      },
    },
    description: 'filter on author: {owner.name}',
    objectID: 'author-filter',
  },
  {
    condition: {
      pattern: 'owner\\: {facet:owner.name}',
      anchoring: 'contains',
    },
    consequence: {
      params: {
        automaticFacetFilters: ['owner.name'],
        query: {
          remove: ['owner\\:', '{facet:owner.name}'],
        },
      },
    },
    description: 'filter on owner: {owner.name}',
    objectID: 'owner-filter',
  },
  {
    condition: {
      pattern: 'keyword\\: {facet:keywords}',
      anchoring: 'contains',
    },
    consequence: {
      params: {
        automaticFacetFilters: ['keywords'],
        query: {
          remove: ['keyword\\:', '{facet:keywords}'],
        },
      },
    },
    description: 'filter on keyword: {keywords}',
    objectID: 'keyword-filter',
  },
];

export const config = {
  npmRegistryEndpoint: 'https://replicate.npmjs.com',
  npmRegistryDBName: 'registry',
  npmDownloadsEndpoint: 'https://api.npmjs.org/downloads',
  npmRootEndpoint: 'https://registry.npmjs.org',
  jsDelivrHitsEndpoint:
    'https://data.jsdelivr.com/v1/stats/packages/npm/month/all',
  jsDelivrPackageEndpoint: 'https://data.jsdelivr.com/v1/package/npm',
  typescriptTypesIndex:
    'https://typespublisher.blob.core.windows.net/typespublisher/data/search-index-min.json',
  maxObjSize: 450000,
  popularDownloadsRatio: 0.005,
  appId: 'OFCNCOG2CU',
  apiKey: '',
  indexName: 'npm-search',
  bootstrapIndexName: 'npm-search-bootstrap',
  bootstrapConcurrency: 25,
  timeToRedoBootstrap: ms('1 month'),
  seq: undefined,
  indexSettings,
  indexSynonyms,
  indexRules,
  expiresAt: ms('30 days'),
  popularExpiresAt: ms('7 days'),
  cacheTotalDownloads: ms('1 minute'),
  prefetchWaitBetweenPage: 5000,
  prefetchMaxIdle: 100,
  retryMax: 2,
  retrySkipped: ms('1 minute'),
  retryBackoffPow: 3,
  refreshPeriod: ms('2 minutes'),
};

export type Config = typeof config;

Object.entries(process.env).forEach(([key, value]) => {
  if (typeof config[key] !== 'undefined') {
    config[key] =
      typeof config[key] === 'number' ? parseInt(value!, 10) : value;
  }
});
