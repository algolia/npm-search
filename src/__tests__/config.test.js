import config from '../config';

it('gets the correct keys from env variables', () => {
  // from mocked .env
  expect(config.apiKey).toBe('fake-api-key');
  // from config.js
  expect(config.maxObjSize).toBe(450000);
  expect(config).toMatchSnapshot();
});
