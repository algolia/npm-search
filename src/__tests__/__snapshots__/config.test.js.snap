// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets the correct keys from env variables 1`] = `
Object {
  "apiKey": "fake-api-key",
  "appId": "fake-app",
  "bootstrapConcurrency": 100,
  "bootstrapIndexName": "npm-search-bootstrap",
  "indexName": "fake-index",
  "indexRules": Array [
    Object {
      "condition": Object {
        "anchoring": "is",
        "pattern": "{facet:concatenatedName}",
      },
      "consequence": Object {
        "params": Object {
          "automaticOptionalFacetFilters": Array [
            "concatenatedName",
          ],
        },
      },
      "description": "promote exact matches",
      "objectID": "promote-exact",
    },
  ],
  "indexSettings": Object {
    "attributesForFaceting": Array [
      "filterOnly(concatenatedName)",
      "searchable(keywords)",
      "searchable(owner.name)",
      "deprecated",
    ],
    "customRanking": Array [
      "desc(downloadsLast30Days)",
    ],
    "disableExactOnAttributes": Array [
      "description",
      "keywords",
      "owner.name",
      "owners.name",
    ],
    "disablePrefixOnAttributes": Array [
      "keywords",
      "owner.name",
      "owners.name",
    ],
    "disableTypoToleranceOnAttributes": Array [
      "keywords",
    ],
    "exactOnSingleWordQuery": "attribute",
    "optionalWords": Array [
      "js",
      "javascript",
    ],
    "ranking": Array [
      "filters",
      "typo",
      "words",
      "proximity",
      "attribute",
      "asc(deprecated)",
      "asc(badPackage)",
      "desc(popular)",
      "exact",
      "custom",
    ],
    "replaceSynonymsInHighlight": false,
    "searchableAttributes": Array [
      "unordered(popularName)",
      "unordered(name)",
      "unordered(concatenatedName)",
      "unordered(description)",
      "unordered(keywords)",
      "owner.name",
      "owners.name",
    ],
    "separatorsToIndex": "_",
  },
  "indexSynonyms": Array [
    Object {
      "objectID": "underscore",
      "synonyms": Array [
        "_",
        "underscore",
      ],
      "type": "synonym",
    },
  ],
  "maxObjSize": 450000,
  "npmDownloadsEndpoint": "https://api.npmjs.org/downloads",
  "npmRegistryEndpoint": "https://replicate.npmjs.com/registry",
  "popularDownloadsRatio": 0.005,
  "replicateConcurrency": 10,
  "seq": null,
  "timeToRedoBootstrap": 604800000,
}
`;
